integration/test/Testlib/App.hs

```diff
@@ -0,0 +1,935 @@
+import Control.Retry (fibonacciBackoff, limitRetriesByCumulativeDelay, retrying)
+import Data.Aeson hiding ((.=))
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as Aeson
+import qualified Data.Aeson.Key as KM
+import qualified Data.Aeson.KeyMap as KM
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as C8
+import Data.ByteString.Conversion
+import qualified Data.ByteString.Lazy as L
+import qualified Data.ByteString.Lazy.Char8 as LC8
+import qualified Data.CaseInsensitive as CI
+import qualified Data.Char as C
+import Data.List.Split (splitOn)
+import qualified Data.Map.Strict as Map
+import Data.Proxy (Proxy (Proxy))
+import qualified Data.Scientific as Sci
+import Data.String.Conversions (cs)
+import Data.Tagged
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Yaml as Yaml
+import GHC.Exception
+import GHC.Records
+import GHC.Stack
+import Imports hiding (ask, asks, local)
+import qualified Network.HTTP.Client as HTTP
+import qualified Network.HTTP.Types as HTTP
+import qualified Network.Socket as N
+import Network.URI (uriToString)
+import System.Exit (exitFailure)
+import System.FilePath (joinPath, takeDirectory, (</>))
+import System.FilePath.Posix (splitPath)
+import System.IO (hPutStrLn, openBinaryTempFile)
+import qualified System.IO.Error as Error
+import System.Process (CreateProcess (..), createProcess, proc, terminateProcess)
+import Test.Tasty.Options
+import Test.Tasty.Providers
+import qualified Test.Tasty.Providers as Tasty
+import Test.Tasty.Providers.ConsoleFormat
+
+-------------------------------------------------------------------------------
+-- - SECTION_APP : App, Env, ServiceMap, Context
+-------------------------------------------------------------------------------
+
+newtype App a = App {unApp :: ReaderT Env IO a}
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadIO,
+      MonadMask,
+      MonadCatch,
+      MonadThrow,
+      MonadReader Env
+    )
+
+data AppFailure = AppFailure String
+
+instance Show AppFailure where
+  show (AppFailure msg) = msg
+
+instance Exception AppFailure where
+  displayException (AppFailure msg) = msg
+
+failApp :: String -> App a
+failApp msg = throw (AppFailure msg)
+
+runAppWithEnv :: Env -> App a -> IO a
+runAppWithEnv e m = runReaderT (unApp m) e
+
+getPrekey :: App Value
+getPrekey = App $ do
+  pks <- asks (.prekeys)
+  (i, pk) <- atomicModifyIORef pks getPK
+  pure $ object ["id" .= i, "key" .= pk]
+  where
+    getPK [] = error "Out of prekeys"
+    getPK (k : ks) = (ks, k)
+
+getLastPrekey :: App Value
+getLastPrekey = App $ do
+  pks <- asks (.lastPrekeys)
+  lpk <- atomicModifyIORef pks getPK
+  pure $ object ["id" .= lastPrekeyId, "key" .= lpk]
+  where
+    getPK [] = error "Out of prekeys"
+    getPK (k : ks) = (ks, k)
+
+    lastPrekeyId :: Int
+    lastPrekeyId = 65535
```

Comment:
these look a bit out of place here...

Reply:
You mean because they are domain-specific? I think it's okay: They modify state in Env, so it makes sense they live in the module that defines the Env.